### ISSUE TITLE
tokio-process stdout+stderr piping deadlocks on Linux

### ISSUE THREAD

**Issue by GeorgeHahn** [tokio-rs/tokio#5526]
**Version**
```
├── tokio v1.26.0
│   └── tokio-macros v1.8.2 (proc-macro)
└── tokio-process v0.2.5
    ├── tokio-io v0.1.13
    ├── tokio-reactor v0.1.12
    │   ├── tokio-executor v0.1.10
    │   ├── tokio-io v0.1.13 (*)
    │   └── tokio-sync v0.1.8
    └── tokio-signal v0.2.9
        ├── tokio-executor v0.1.10 (*)
        ├── tokio-io v0.1.13 (*)
        └── tokio-reactor v0.1.12 (*)
```

**Platform**
`Linux <hostname> 5.10.0-15-cloud-amd64 #1 SMP Debian 5.10.120-1 (2022-06-09) x86_64 GNU/Linux`

**Description**
Reading piped stdout & stderr from a tokio-process command hangs on Linux. A reproduction is included below; this reproduces 100% of the time in debug and release builds for me.

Reproducing code: 
```rust
use std::process::Stdio;
use tokio::{fs::File, io::BufReader, process::Command};

#[tokio::main]
async fn main() {
    let mut stdout_log = File::create("stdout.log").await.unwrap();
    let mut stderr_log = File::create("stderr.log").await.unwrap();

    let mut child = Command::new("target/release/child")
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .expect("Could not spawn cmd");

    let mut stdout = BufReader::new(child.stdout.take().unwrap());
    tokio::io::copy(&mut stdout, &mut stdout_log).await.unwrap();

    let mut stderr = BufReader::new(child.stderr.take().unwrap());
    tokio::io::copy(&mut stderr, &mut stderr_log).await.unwrap();

    child.wait().await.unwrap();
}
```

Example child:
```rust
fn main() {
    for _ in 0..100 {
        for _ in 0..100 {
            print!("words, ");
            eprint!("errors, ");
        }
        println!("Words!");
        eprintln!("Errors!");
    }
}
```

Full project: [tokio-process-minrepr.tar.gz](https://github.com/tokio-rs/tokio/files/10883426/tokio-process-minrepr.tar.gz)
Run `cargo build --release && ./target/release/minrepr`

All threads are parked, except for one waiting on epoll:
```
Thread 2 (Thread 0x7efcf0a3a700 (LWP 1237063) "tokio-runtime-w"):
#0  0x00007efcf0b39d56 in epoll_wait (epfd=3, events=0x55c902c72d10, maxevents=1024, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
#1  0x000055c901c054d8 in mio::sys::unix::selector::epoll::Selector::select (self=0x55c902c7bb20, events=0x55c902c7bb30, timeout=...) at src/sys/unix/selector/epoll.rs:101
#2  0x000055c901c0477d in mio::poll::Poll::poll (self=0x55c902c7bb20, events=0x55c902c7bb30, timeout=...) at src/poll.rs:411
#3  0x000055c901b8aa57 in tokio::runtime::io::Driver::turn (self=0x55c902c7bb20, handle=0x55c902c84ab8, max_wait=...) at src/runtime/io/mod.rs:172
#4  0x000055c901b8a8ae in tokio::runtime::io::Driver::park (self=0x55c902c7bb20, rt_handle=0x55c902c84ab8) at src/runtime/io/mod.rs:138
#5  0x000055c901bab85b in tokio::runtime::signal::Driver::park (self=0x55c902c7bb08, handle=0x55c902c84ab8) at src/runtime/signal/mod.rs:92
#6  0x000055c901bab3d7 in tokio::runtime::process::Driver::park (self=0x55c902c7bb08, handle=0x55c902c84ab8) at src/runtime/process.rs:32
#7  0x000055c901b87d44 in tokio::runtime::driver::IoStack::park (self=0x55c902c7bb08, handle=0x55c902c84ab8) at src/runtime/driver.rs:159
#8  0x000055c901bc7527 in tokio::runtime::time::Driver::park_internal (self=0x55c902c7bb08, rt_handle=0x55c902c84ab8, limit=...) at src/runtime/time/mod.rs:211
#9  0x000055c901bc6ff4 in tokio::runtime::time::Driver::park (self=0x55c902c7bb08, handle=0x55c902c84ab8) at src/runtime/time/mod.rs:152
#10 0x000055c901b88336 in tokio::runtime::driver::TimeDriver::park (self=0x55c902c7bb00, handle=0x55c902c84ab8) at src/runtime/driver.rs:303
#11 0x000055c901b87523 in tokio::runtime::driver::Driver::park (self=0x55c902c7bb00, handle=0x55c902c84ab8) at src/runtime/driver.rs:62
#12 0x000055c901bc5eae in tokio::runtime::scheduler::multi_thread::park::Inner::park_driver (self=0x55c902c72b10, driver=0x55c902c7bb00, handle=0x55c902c84ab8) at src/runtime/scheduler/multi_thread/park.rs:189
#13 0x000055c901bc59d3 in tokio::runtime::scheduler::multi_thread::park::Inner::park (self=0x55c902c72b10, handle=0x55c902c84ab8) at src/runtime/scheduler/multi_thread/park.rs:122
#14 0x000055c901bc5514 in tokio::runtime::scheduler::multi_thread::park::Parker::park (self=0x7efcf0a38c70, handle=0x55c902c84ab8) at src/runtime/scheduler/multi_thread/park.rs:68
#15 0x000055c901b51fd8 in tokio::runtime::scheduler::multi_thread::worker::Context::park_timeout (self=0x7efcf0a39060, core=0x55c902c72aa0, duration=...) at src/runtime/scheduler/multi_thread/worker.rs:560
#16 0x000055c901b51b0d in tokio::runtime::scheduler::multi_thread::worker::Context::park (self=0x7efcf0a39060, core=0x55c902c72aa0) at src/runtime/scheduler/multi_thread/worker.rs:531
#17 0x000055c901b508bc in tokio::runtime::scheduler::multi_thread::worker::Context::run (self=0x7efcf0a39060, core=0x55c902c72aa0) at src/runtime/scheduler/multi_thread/worker.rs:439
#18 0x000055c901b5053d in tokio::runtime::scheduler::multi_thread::worker::run::{closure#0} () at src/runtime/scheduler/multi_thread/worker.rs:406
#19 0x000055c901b5ac47 in tokio::macros::scoped_tls::ScopedKey<tokio::runtime::scheduler::multi_thread::worker::Context>::set<tokio::runtime::scheduler::multi_thread::worker::Context, tokio::runtime::scheduler::multi_thread::worker::run::{closure_env#0}, ()> (self=0x55c901cbdd50 <tokio::runtime::scheduler::multi_thread::worker::CURRENT>, t=0x7efcf0a39060, f=...) at src/macros/scoped_tls.rs:61
#20 0x000055c901b5039f in tokio::runtime::scheduler::multi_thread::worker::run (worker=...) at src/runtime/scheduler/multi_thread/worker.rs:403
#21 0x000055c901b5013b in tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure#0} () at src/runtime/scheduler/multi_thread/worker.rs:365
#22 0x000055c901bba38e in tokio::runtime::blocking::task::{impl#2}::poll<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}, ()> (self=..., _cx=0x7efcf0a39228) at src/runtime/blocking/task.rs:42
#23 0x000055c901b55180 in tokio::runtime::task::core::{impl#6}::poll::{closure#0}<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule> (ptr=0x55c902c85028) at src/runtime/task/core.rs:223
#24 0x000055c901b85b0f in tokio::loom::std::unsafe_cell::UnsafeCell<tokio::runtime::task::core::Stage<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>>>::with_mut<tokio::runtime::task::core::Stage<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>>, core::task::poll::Poll<()>, tokio::runtime::task::core::{impl#6}::poll::{closure_env#0}<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule>> (self=0x55c902c85028, f=...) at src/loom/std/unsafe_cell.rs:14
#25 0x000055c901b54e2e in tokio::runtime::task::core::Core<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule>::poll<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule> (self=0x55c902c85020, cx=...) at src/runtime/task/core.rs:212
#26 0x000055c901b71cf8 in tokio::runtime::task::harness::poll_future::{closure#0}<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule> () at src/runtime/task/harness.rs:476
#27 0x000055c901ba25c3 in core::panic::unwind_safe::{impl#23}::call_once<core::task::poll::Poll<()>, tokio::runtime::task::harness::poll_future::{closure_env#0}<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule>> (self=..., _args=()) at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/panic/unwind_safe.rs:271
#28 0x000055c901bd3fdd in std::panicking::try::do_call<core::panic::unwind_safe::AssertUnwindSafe<tokio::runtime::task::harness::poll_future::{closure_env#0}<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule>>, core::task::poll::Poll<()>> (data=0x7efcf0a39390) at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:483
#29 0x000055c901bd4aab in __rust_try ()
#30 0x000055c901bd2681 in std::panicking::try<core::task::poll::Poll<()>, core::panic::unwind_safe::AssertUnwindSafe<tokio::runtime::task::harness::poll_future::{closure_env#0}<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule>>> (f=...) at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:447
#31 0x000055c901b694db in std::panic::catch_unwind<core::panic::unwind_safe::AssertUnwindSafe<tokio::runtime::task::harness::poll_future::{closure_env#0}<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule>>, core::task::poll::Poll<()>> (f=...) at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panic.rs:137
#32 0x000055c901b71334 in tokio::runtime::task::harness::poll_future<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule> (core=0x55c902c85020, cx=...) at src/runtime/task/harness.rs:464
#33 0x000055c901b6ef29 in tokio::runtime::task::harness::Harness<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule>::poll_inner<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule> (self=0x7efcf0a395c0) at src/runtime/task/harness.rs:198
#34 0x000055c901b6edf7 in tokio::runtime::task::harness::Harness<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule>::poll<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule> (self=...) at src/runtime/task/harness.rs:152
#35 0x000055c901bb9d5d in tokio::runtime::task::raw::poll<tokio::runtime::blocking::task::BlockingTask<tokio::runtime::scheduler::multi_thread::worker::{impl#0}::launch::{closure_env#0}>, tokio::runtime::blocking::schedule::BlockingSchedule> (ptr=...) at src/runtime/task/raw.rs:255
#36 0x000055c901bb9bc8 in tokio::runtime::task::raw::RawTask::poll (self=...) at src/runtime/task/raw.rs:200
#37 0x000055c901bc7fc7 in tokio::runtime::task::UnownedTask<tokio::runtime::blocking::schedule::BlockingSchedule>::run<tokio::runtime::blocking::schedule::BlockingSchedule> (self=...) at src/runtime/task/mod.rs:431
#38 0x000055c901b5b837 in tokio::runtime::blocking::pool::Task::run (self=...) at src/runtime/blocking/pool.rs:159
#39 0x000055c901b5ecae in tokio::runtime::blocking::pool::Inner::run (self=0x55c902c7b9e0, worker_thread_id=0) at src/runtime/blocking/pool.rs:511
#40 0x000055c901b5e1e4 in tokio::runtime::blocking::pool::{impl#6}::spawn_thread::{closure#0} () at src/runtime/blocking/pool.rs:469
#41 0x000055c901b7f6f1 in std::sys_common::backtrace::__rust_begin_short_backtrace<tokio::runtime::blocking::pool::{impl#6}::spawn_thread::{closure_env#0}, ()> (f=...) at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/sys_common/backtrace.rs:121
#42 0x000055c901b61162 in std::thread::{impl#0}::spawn_unchecked_::{closure#1}::{closure#0}<tokio::runtime::blocking::pool::{impl#6}::spawn_thread::{closure_env#0}, ()> () at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/thread/mod.rs:550
#43 0x000055c901ba2642 in core::panic::unwind_safe::{impl#23}::call_once<(), std::thread::{impl#0}::spawn_unchecked_::{closure#1}::{closure_env#0}<tokio::runtime::blocking::pool::{impl#6}::spawn_thread::{closure_env#0}, ()>> (self=..., _args=()) at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/panic/unwind_safe.rs:271
#44 0x000055c901bd3a4e in std::panicking::try::do_call<core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked_::{closure#1}::{closure_env#0}<tokio::runtime::blocking::pool::{impl#6}::spawn_thread::{closure_env#0}, ()>>, ()> (data=0x7efcf0a39b00) at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:483
#45 0x000055c901bd4aab in __rust_try ()
#46 0x000055c901bd2523 in std::panicking::try<(), core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked_::{closure#1}::{closure_env#0}<tokio::runtime::blocking::pool::{impl#6}::spawn_thread::{closure_env#0}, ()>>> (f=...) at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:447
#47 0x000055c901b695f0 in std::panic::catch_unwind<core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked_::{closure#1}::{closure_env#0}<tokio::runtime::blocking::pool::{impl#6}::spawn_thread::{closure_env#0}, ()>>, ()> (f=...) at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panic.rs:137
#48 0x000055c901b60eef in std::thread::{impl#0}::spawn_unchecked_::{closure#1}<tokio::runtime::blocking::pool::{impl#6}::spawn_thread::{closure_env#0}, ()> () at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/thread/mod.rs:549
#49 0x000055c901b930af in core::ops::function::FnOnce::call_once<std::thread::{impl#0}::spawn_unchecked_::{closure_env#1}<tokio::runtime::blocking::pool::{impl#6}::spawn_thread::{closure_env#0}, ()>, ()> () at /rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/ops/function.rs:507
#50 0x000055c901c47293 in alloc::boxed::{impl#45}::call_once<(), dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global> () at library/alloc/src/boxed.rs:2000
#51 alloc::boxed::{impl#45}::call_once<(), alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global>, alloc::alloc::Global> () at library/alloc/src/boxed.rs:2000
#52 std::sys::unix::thread::{impl#2}::new::thread_start () at library/std/src/sys/unix/thread.rs:108
#53 0x00007efcf0d63ea7 in start_thread (arg=<optimized out>) at pthread_create.c:477
#54 0x00007efcf0b39a2f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

**Comment by GeorgeHahn** [comment]
Wow, huge miss on my side. Thank you!!
